File: JavaLibrary/src/main/java/com/flat/simplex/lexer/Key.java
package com.flat.simplex.lexer;

import java.util.HashMap;

public enum Key {
    Invalid(""),

    Word(""),
    String(""),
    Number(""),

    If("if"),
    Else("else"),
    Switch("switch"),
    Case("case"),
    Default("default"),
    While("while"),
    For("for"),
    Do("do"),
    Break("break"),
    Continue("continue"),
    Return("return"),
    With("with"),
    Var("var"),

    Function("function"),
    Undefined("undefined"),
    True("true"),
    False("false"),
    Global("global"),
    This("this"),

    Brace("{"),
    CBrace("}"),
    Param("("),
    CParam(")"),
    Index("["),
    CIndex("]"),

    Dot("."),
    Comma(","),
    Semicolon(";"),
    Colon(":", 12),
    Quest("?", 12),

    Add("+", 3),
    Sub("-", 3),
    Mul("*", 2),
    Div("/", 2),
    Mod("%", 2),
    Sfl("<<", 4),
    Sfr(">>", 4),

    Inc("++", 1),
    Dec("--", 1),
    Not("!", 1),
    Bnot("~", 1),

    Set("=", 13),
    Setadd("+=", 13),
    Setsub("-=", 13),
    Setmul("*=", 13),
    Setdiv("/=", 13),
    Setmod("%=", 13),
    Setsfl("<<=", 13),
    Setsfr(">>=", 13),

    Eq("==", 6),
    Dif("!=", 6),
    Gr(">", 5),
    Gre(">=", 5),
    Ls("<", 5),
    Lse("<=", 5),

    Band("&", 7),
    Bxor("^", 8),
    Bor("|", 9),

    And("&&", 10),
    Or("||", 11),

    Hex("#");

    public final String name;
    public final int op;
    private static HashMap<Integer, Key> keys;

    Key(String name) {
        this.name = name;
        this.op = 0;
    }

    Key(String name, int operatorType) {
        this.name = name;
        this.op = operatorType;
    }

    /**
     * Try to find a key from a source.
     *
     * @param source String source
     * @param start Start index offset
     * @param end End index offset
     * @return The Key found, or null
     */
    public static Key readKey(String source, int start, int end) {
        if (keys == null) {
            keys = new HashMap<>();
            for (var keyWord : values()) {
                if (keyWord.name.length() > 0) keys.put(keyWord.name.hashCode(), keyWord);
            }
        }

        int h = 0;
        int length = end - start;
        if (length <= 0 || length > 9) return null;

        for (int i = start; i < end; i++) {
            h = 31 * h + source.charAt(i);
        }
        Key key = keys.get(h);
        return key != null && isEqual(key, source, start, end) ? key : null;
    }

    private static boolean isEqual(Key key, String source, int start, int end) {
        int len = end - start;
        if (len == key.name.length()) {
            for (int i = 0; i < len; i++) {
                if (source.charAt(i + start) != key.name.charAt(i)) {
                    return false;
                }
            }
            return true;
        } else {
            return false;
        }
    }

}

File: SimplexApp/simplex/lexer/Key.js
class Key {
    constructor (name, operatorType) {
        this.name = name;
        this.op = operatorType ? operatorType : 0;
    }

    static hashCode(str, start, end) {
        if (!start) start = 0;
        if (!end) end = str.length;

        let hash = 0;
        if (str.length === 0) return hash;
        for (let i = start; i < end; i++) {
            hash = ((hash << 5) - hash) + str.charCodeAt(i);
            hash = hash & hash;
        }
        return hash;
    }

    static readKey(source, start, end) {
        if (Key.keys === null) {
            Key.keys = new Map();
            for (const [name, value] of Object.entries(Key)) {
                if (value instanceof Key && name.length > 0) {
                    Key.keys[Key.hashCode(value.name)] = value;
                }
            }
        }
        let key = Key.keys[Key.hashCode(source, start, end)];
        return key ? key : null;
    }
}

Key.Invalid = new Key('');
Key.Word = new Key('');
Key.String = new Key('');
Key.Number = new Key('');
Key.If = new Key('if');
Key.Else = new Key('else');
Key.Switch = new Key('switch');
Key.Case = new Key('case');
Key.Default = new Key('default');
Key.While = new Key('while');
Key.For = new Key('for');
Key.Do = new Key('do');
Key.Break = new Key('break');
Key.Continue = new Key('continue');
Key.Return = new Key('return');
Key.With = new Key('with');
Key.Var = new Key('var');
Key.Function = new Key('function');
Key.Undefined = new Key('undefined');
Key.True = new Key('true');
Key.False = new Key('false');
Key.Global = new Key('global');
Key.This = new Key('this');
Key.Brace = new Key('{');
Key.CBrace = new Key('}');
Key.Param = new Key('(');
Key.CParam = new Key(')');
Key.Index = new Key('[');
Key.CIndex = new Key(']');
Key.Dot = new Key('.');
Key.Comma = new Key(',');
Key.Semicolon = new Key(';');
Key.Colon = new Key(':', 12);
Key.Quest = new Key('?', 12);
Key.Add = new Key('+', 3);
Key.Sub = new Key('-', 3);
Key.Mul = new Key('*', 2);
Key.Div = new Key('/', 2);
Key.Mod = new Key('%', 2);
Key.Sfl = new Key('<<', 4);
Key.Sfr = new Key('>>', 4);
Key.Inc = new Key('++', 1);
Key.Dec = new Key('--', 1);
Key.Not = new Key('!', 1);
Key.Bnot = new Key('~', 1);
Key.Set = new Key('=', 13);
Key.Setadd = new Key('+=', 13);
Key.Setsub = new Key('-=', 13);
Key.Setmul = new Key('*=', 13);
Key.Setdiv = new Key('/=', 13);
Key.Setmod = new Key('%=', 13);
Key.Setsfl = new Key('<<=', 13);
Key.Setsfr = new Key('>>=', 13);
Key.Eq = new Key('==', 6);
Key.Dif = new Key('!=', 6);
Key.Gr = new Key('>', 5);
Key.Gre = new Key('>=', 5);
Key.Ls = new Key('<', 5);
Key.Lse = new Key('<=', 5);
Key.Band = new Key('&', 7);
Key.Bxor = new Key('^', 8);
Key.Bor = new Key('|', 9);
Key.And = new Key('&&', 10);
Key.Or = new Key('||', 11);
Key.Hex = new Key('#');
Key.keys = null;

module.exports = Key;